flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -367,53 +375,299 @@ SqlDrop SqlDropView(Span s, boolean replace) :
     }
 }
 
+SqlIdentifier FlinkCollectionsTypeName() :
+{
+}
+{
+    LOOKAHEAD(2)
+    <MULTISET> {
+        return new SqlIdentifier(SqlTypeName.MULTISET.name(), getPos());
+    }
+|
+    <ARRAY> {
+        return new SqlIdentifier(SqlTypeName.ARRAY.name(), getPos());
+    }
+}
+
+SqlIdentifier FlinkTypeName() :
+{
+    final SqlTypeName sqlTypeName;
+    final SqlIdentifier typeName;
+    final Span s = Span.of();
+}
+{
+    (
+<#-- additional types are included here -->
+<#-- make custom data types in front of Calcite core data types -->
+<#list parser.flinkDataTypeParserMethods as method>
+    <#if (method?index > 0)>
+    |
+    </#if>
+        LOOKAHEAD(2)
+        typeName = ${method}
+</#list>
+    |
+        LOOKAHEAD(2)
+        sqlTypeName = SqlTypeName(s) {
+            typeName = new SqlIdentifier(sqlTypeName.name(), s.end(this));
+        }
+    |
+        LOOKAHEAD(2)
+        typeName = FlinkCollectionsTypeName()
+    |
+        typeName = CompoundIdentifier() {
+            throw new ParseException("UDT in DDL is not supported yet.");
```

Comment:
But this means that the parser is still not compatible with FLIP-37. So it is still not possible to execute `CAST(12 AS STRING)`?

Reply:
Yes. Let's postpone full type support for cast for 1.10, when i finished the data type parse refactoring in Calcite[1], which would be into release 1.21.

BTW, the full implicit type cast[2] support would be into release 1.21 too.

[1] https://issues.apache.org/jira/browse/CALCITE-3213
[2] https://issues.apache.org/jira/browse/CALCITE-2302

Reply:
Agreed, but please open an issue also in Flink to track the support. It is very confusing for users if data types are only available in certain SQL locations. Users don't understand the difference between DDL and DQL. Esp. with the introduction of computed columns it gets more confusing because special cases like `STRING` would be available in DDL but not in the `CAST` of the computed column.